File: firebase/functions/lib/events/live_meetings/get_meeting_poll_data.dart
import 'dart:async';

import 'package:data_models/user_input/poll_data.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart'
    as functions_interop;
import '../../on_call_function.dart';
import '../../utils/infra/firebase_auth_utils.dart';
import '../../utils/infra/firestore_utils.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/events/live_meetings/meeting_guide.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/user/public_user_info.dart';

class GetMeetingPollData extends OnCallMethod<GetMeetingPollDataRequest> {
  GetMeetingPollData()
      : super(
          'GetMeetingPollData',
          (jsonMap) => GetMeetingPollDataRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetMeetingPollDataRequest request,
    functions_interop.CallableContext context,
  ) async {
    final eventPath = request.eventPath;
    final match = RegExp('/?community/([^/]+)/templates/([^/]+)/events/([^/]+)')
        .matchAsPrefix(eventPath);
    if (match == null) {
      throw functions_interop.HttpsError(
        functions_interop.HttpsError.invalidArgument,
        'Path malformed.',
        null,
      );
    }
    final communityId = match.group(1);
    final eventId = match.group(3);

    final event = await firestoreUtils.getFirestoreObject(
      path: eventPath,
      constructor: (map) => Event.fromJson(map),
    );

    final membershipDoc =
        'memberships/${context.authUid}/community-membership/$communityId';
    final communityMembershipDoc =
        await firestore.document(membershipDoc).get();
    final membership = Membership.fromJson(
      firestoreUtils.fromFirestoreJson(communityMembershipDoc.data.toMap()),
    );

    orElseUnauthorized(
      membership.isAdmin || event.creatorId == context.authUid,
    );

    final liveMeetingPath = '${request.eventPath}/live-meetings/$eventId';
    final breakoutRoomSessions = '$liveMeetingPath/breakout-room-sessions';
    print(breakoutRoomSessions);
    final sessionDocs = await firestore.collection(breakoutRoomSessions).get();
    final roomDocQueries = await Future.wait(
      sessionDocs.documents.map((session) {
        final collectionPath = '${session.reference.path}/breakout-rooms';
        print(collectionPath);
        return firestore.collection(collectionPath).get();
      }),
    );
    final roomDocs =
        roomDocQueries.map((query) => query.documents).expand((a) => a);
    final breakoutMeetingLinks = roomDocs.map(
      (roomDoc) =>
          '${roomDoc.reference.path}/live-meetings/${roomDoc.documentID}',
    );

    final meetingPaths = [
      request.eventPath,
      ...breakoutMeetingLinks,
    ];

    final pollDataListResults = await Future.wait(<Future<List<PollData>>>[
      for (final path in meetingPaths) _getPollsFromPath(path, event),
    ]);

    return GetMeetingPollDataResponse(
      polls: pollDataListResults.expand((p) => p).toList(),
    ).toJson();
  }

  Future<List<PollData>> _getPollsFromPath(String path, Event event) async {
    final roomId = path.split('/').last;

    final pollAgendaItems =
        event.agendaItems.where((a) => a.type == AgendaItemType.poll).toList();

    final pollDataList = <PollData>[];

    for (final agendaItem in pollAgendaItems) {
      final participantDetailsPath =
          '$path/participant-agenda-item-details/${agendaItem.id}/participant-details';

      final documents =
          await firestore.collection(participantDetailsPath).get();

      for (var document in documents.documents) {
        final details = ParticipantAgendaItemDetails.fromJson(
          firestoreUtils.fromFirestoreJson(document.data.toMap()),
        );

        if (details.pollResponse != null) {
          final memberInfo = await firebaseAuthUtils.getUser(details.userId!);

          String memberName;
          if (isNullOrEmpty(memberInfo.displayName)) {
            final memberDoc =
                await firestore.document('publicUser/${memberInfo.uid}').get();
            final publicUserInfo = PublicUserInfo.fromJson(
              firestoreUtils.fromFirestoreJson(memberDoc.data.toMap()),
            );
            memberName = publicUserInfo.displayName ?? '';
          } else {
            memberName = memberInfo.displayName;
          }

          final pollData = PollData(
            userId: details.userId,
            userName: memberName,
            userEmail: memberInfo.email,
            agendaItemId: agendaItem.id,
            pollQuestion: agendaItem.title,
            pollResponse: details.pollResponse,
            roomId: roomId,
            answeredDate: document.updateTime?.toDateTime(),
          );

          pollDataList.add(pollData);
        }
      }
    }

    return pollDataList;
  }
}

File: firebase/functions/src/events/live_meetings/get_meeting_poll_data.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { firebaseAuthUtils } from '../../utils/infra/firebase_auth_utils';
import { orElseUnauthorized } from '../../utils/utils';
import { Event, Membership, Participant, ParticipantStatus, AgendaItemType } from '../../types';

interface GetMeetingPollDataRequest {
  eventPath: string;
}

interface PollData {
  userId?: string;
  userName?: string;
  userEmail?: string;
  agendaItemId?: string;
  pollQuestion?: string;
  pollResponse?: unknown;
  roomId?: string;
  answeredDate?: Date;
}

export class GetMeetingPollData extends OnCallMethod<GetMeetingPollDataRequest> {
  constructor() {
    super('GetMeetingPollData', (jsonMap) => jsonMap as GetMeetingPollDataRequest);
  }

  async action(request: GetMeetingPollDataRequest, context: functions.https.CallableContext): Promise<Record<string, unknown>> {
    const match = /\/?community\/([^/]+)\/templates\/([^/]+)\/events\/([^/]+)/.exec(request.eventPath);
    if (!match) throw new functions.https.HttpsError('invalid-argument', 'Path malformed.');
    const communityId = match[1];
    const eventId = match[3];

    const event: Event = await firestoreUtils.getFirestoreObject({
      path: request.eventPath,
      constructor: (map) => map as unknown as Event,
    });

    const membershipSnap = await firestore.doc(`memberships/${context.auth?.uid}/community-membership/${communityId}`).get();
    const membership: Membership = firestoreUtils.fromFirestoreJson(membershipSnap.data() ?? {}) as unknown as Membership;
    orElseUnauthorized((membership as any).isAdmin || event.creatorId === context.auth?.uid);

    const liveMeetingPath = `${request.eventPath}/live-meetings/${eventId}`;
    const breakoutRoomSessions = `${liveMeetingPath}/breakout-room-sessions`;
    const sessionDocs = await firestore.collection(breakoutRoomSessions).get();

    const roomDocQueries = await Promise.all(
      sessionDocs.docs.map((session) => firestore.collection(`${session.ref.path}/breakout-rooms`).get())
    );
    const roomDocs = roomDocQueries.flatMap((q) => q.docs);
    const breakoutMeetingLinks = roomDocs.map((roomDoc) => `${roomDoc.ref.path}/live-meetings/${roomDoc.id}`);
    const meetingPaths = [liveMeetingPath, ...breakoutMeetingLinks];

    const pollDataListResults = await Promise.all(meetingPaths.map((path) => this._getPollsFromPath(path, event)));
    const allPolls = pollDataListResults.flat();

    return { polls: allPolls };
  }

  private async _getPollsFromPath(path: string, event: Event): Promise<PollData[]> {
    const roomId = path.split('/').pop()!;
    const pollAgendaItems = (event.agendaItems ?? []).filter((a: any) => a.type === AgendaItemType?.poll || a.type === 'poll');
    const pollDataList: PollData[] = [];

    for (const agendaItem of pollAgendaItems) {
      const participantDetailsPath = `${path}/participant-agenda-item-details/${agendaItem.id}/participant-details`;
      const docs = await firestore.collection(participantDetailsPath).get();
      for (const doc of docs.docs) {
        const details = firestoreUtils.fromFirestoreJson(doc.data()) as unknown as any;
        if (details.pollResponse != null) {
          const memberInfo = await firebaseAuthUtils.getUser(details.userId);
          const memberDoc = await firestore.doc(`publicUser/${details.userId}`).get();
          const memberName = memberDoc.data()?.displayName ?? '';
          pollDataList.push({
            userId: details.userId,
            userName: memberName,
            userEmail: memberInfo?.email,
            agendaItemId: agendaItem.id,
            pollQuestion: agendaItem.content,
            pollResponse: details.pollResponse,
            roomId,
            answeredDate: doc.updateTime?.toDate(),
          });
        }
      }
    }

    return pollDataList;
  }
}
